spring/spring6/spring6-common/src/main/java/org/infinispan/spring/common/provider/SpringCache.java

```diff
@@ -259,4 +268,59 @@ static RuntimeException throwValueRetrievalException(Object key, Callable<?> loa
    public long getWriteTimeout() {
       return writeTimeout;
    }
+
+   @Override
+   public CompletableFuture<?> retrieve(Object key) {
+      return encodedToValueWrapper(nativeCache.getAsync(key));
+   }
+
+   @Override
+   public <T> CompletableFuture<T> retrieve(Object key, Supplier<CompletableFuture<T>> valueLoaderAsync) {
+      CompletionStage<T> completionStage = CompletionStages.handleAndCompose(nativeCache.getAsync(key), (v1, ex1) -> {
+         if (ex1 != null) {
+            return CompletableFuture.failedFuture(ex1);
+         }
+
+         if (v1 != null) {
+            return CompletableFuture.completedFuture(decodeNull(v1));
+         }
+
+         CompletableFuture<T> completableFuture = computationResults.get(key);
+         if (completableFuture != null) {
+            return completableFuture;
+         }
+         computationResults.putIfAbsent(key, new CompletableFuture<>());
```

Comment:
The get here isn't needed and should just be tested from the return of putIfAbsent.